File: core/core-api/src/main/java/com/coffee/gu/stamp/StampReader.java
package com.coffee.gu.stamp;

import com.coffee.gu.Principal;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;

@Component
public class StampReader {
    private final StampRepository stampRepository;
    private final StampHistoryRepository stampHistoryRepository;

    public StampReader(StampRepository stampRepository, StampHistoryRepository stampHistoryRepository) {
        this.stampRepository = stampRepository;
        this.stampHistoryRepository = stampHistoryRepository;
    }

    public Long getAvailableStampCounts(Principal principal) {
        return stampRepository.countAvailableStamps(principal.getKey(), LocalDateTime.now());
    }

    public Long getExpiringSoonCount(Principal principal, LocalDateTime now) {
        return stampRepository.countExpiringStamps(principal.getKey(), now, now.plusDays(Stamp.EXPIRY_ALARM_DAYS));
    }

    public List<StampHistory> getStampHistories(Principal principal) {
        return stampHistoryRepository.getWithinNMonths(principal.getKey(), LocalDateTime.now().minusMonths(Stamp.EXPIRY_DAYS));
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/stamp/StampReader.kt
package com.coffee.gu.stamp

import com.coffee.gu.Principal
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class StampReader(
    private val stampRepository: StampRepository,
    private val stampHistoryRepository: StampHistoryRepository,
) {
    fun getAvailableStampCounts(principal: Principal): Int {
        return stampRepository.countAvailableStamps(principal.key, LocalDateTime.now()).toInt()
    }

    fun getExpiringSoonCount(principal: Principal, now: LocalDateTime): Int {
        return stampRepository.countExpiringStamps(principal.key, now, now.plusDays(Stamp.EXPIRY_ALARM_DAYS)).toInt()
    }

    fun getStampHistories(principal: Principal): List<StampHistory> {
        return stampHistoryRepository.getWithinNMonths(principal.key, LocalDateTime.now().minusMonths(Stamp.EXPIRY_DAYS))
    }
}
